### ISSUE TITLE
Use observers for rare events

### ISSUE THREAD

**Issue by cBournhonesque** [cBournhonesque/lightyear#517]
Currently there are a lot of systems that we run that listen on rare events, such as ClientDisconnected.
These systems are scheduled every frame even though the events are actually very rare.

We should instead use observers to only run the logic when the event actually happens.

**Comment by Nul-led** [comment]
What about messages, they can either be rare or not, letting the user choose would prob be hard, no?

**Comment by cBournhonesque** [comment]
Yes it might be worth exploring that for receiving messages as well.
I think it's probably best to wait a bit until improvements on observer/event ergonomics are made, such as https://github.com/bevyengine/bevy/pull/13953
